File: components/subresource_filter/android/java/src/org/chromium/components/subresource_filter/SubresourceFilterFeatureMap.java
// Copyright 2023 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.subresource_filter;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.base.FeatureMap;

/**
 * Java accessor for subresource_filter base::Features.
 *
 * Note: Features must be added to the array |kFeaturesExposedToJava| in
 * //components/subresource_filter/core/browser/subresource_filter_feature_map.cc.
 */
@JNINamespace("subresource_filter")
public final class SubresourceFilterFeatureMap extends FeatureMap {
    // Features exposed through this FeatureMap
    public static final String SUBRESOURCE_FILTER = "SubresourceFilter";

    private static final SubresourceFilterFeatureMap sInstance = new SubresourceFilterFeatureMap();

    // Do not instantiate this class.
    private SubresourceFilterFeatureMap() {}

    /** @return the singleton SubresourceFilterFeatureMap. */
    public static SubresourceFilterFeatureMap getInstance() {
        return sInstance;
    }

    /** Convenience method to call {@link #isEnabledInNative(String)} statically. */
    public static boolean isEnabled(String featureName) {
        return getInstance().isEnabledInNative(featureName);
    }

    /** @return whether SubresourceFilter is enabled. */
    public static boolean isSubresourceFilterEnabled() {
        return isEnabled(SUBRESOURCE_FILTER);
    }

    @Override
    protected long getNativeMap() {
        return SubresourceFilterFeatureMapJni.get().getNativeMap();
    }

    @NativeMethods
    public interface Natives {
        long getNativeMap();
    }
}

File: components/subresource_filter/android/subresource_filter_feature_map.cc
// Copyright 2023 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/android/feature_map.h"
#include "base/feature_list.h"
#include "base/no_destructor.h"
#include "components/subresource_filter/core/browser/subresource_filter_features.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "components/subresource_filter/android/subresource_filter_jni_headers/SubresourceFilterFeatureMap_jni.h"

namespace subresource_filter {

namespace {

// Array of features exposed through the Java SubresourceFilterFeatureMap API.
// Entries in this array may refer to features either defined in
// components/subresource_filter/core/browser/subresource_filter_features.h or
// in other locations in the code base (e.g. content_features.h).
const base::Feature* const kFeaturesExposedToJava[] = {
    &kSafeBrowsingSubresourceFilter,
};

// static
base::android::FeatureMap* GetFeatureMap() {
  static base::NoDestructor<base::android::FeatureMap> kFeatureMap(std::vector(
      std::begin(kFeaturesExposedToJava), std::end(kFeaturesExposedToJava)));
  return kFeatureMap.get();
}

}  // namespace

static jlong JNI_SubresourceFilterFeatureMap_GetNativeMap(JNIEnv* env) {
  return reinterpret_cast<jlong>(GetFeatureMap());
}

}  // namespace subresource_filter
